### ISSUE TITLE
File Operations Slow, Buggy, and Inconsistent

### ISSUE THREAD

**Issue by horsburgh** [hydroshare/hydroshare#2266]
There have now been multiple issues reported by multiple users about weird things happening with file uploads to resources. See #2253, #2224, and #2226. For relatively small numbers of files, the drag and drop upload seems to work OK. But, with larger numbers of files, the performance seems to be unpredictable. This also includes the unzipping procedure, which I have found to also be unpredictable. I've not been able to fully test this because each operation is so slow, but I did some work tonight using this resource and testing in Chrome on my Mac:

https://www.hydroshare.org/resource/11655ab5cc584f54bd6c2c7223cc64d3/

In this resource, there is (was) an 8.6 MB zip file that contains a single folder that has inside of it 478 individual CSV files, each of which is on the order of 65 KB.  So, while the number of files is relatively large, this is not a large volume of data.

Here's some things that I have observed. These all likely need to be tested more systematically and perhaps divided up into multiple issues to fix.

1. The zip file uploads to a new resource fine. No problems there.
2. Unzipping the file takes multiple minutes. This seems really slow. But, the first time I tried it, it did work.
3. Once unzipped, it takes many seconds to minutes to open the folder of files created by the unzip operation just to view its contents.
4. It takes many seconds to minutes before I can initiate a download of any of the files (e.g., this is the time that elapses with a spinning wheel from the time that I right click and choose download file to when the download actually happens). The actual download itself appears to be happening quickly. I'm not sure what is happening in the time leading up to the actual download.
5. Deletion of the folder resulted in a "Bad Gateway" error of some sort. After getting this error, it appeared as though the folder had been removed, but I don't think it was done cleanly, and this is where the real issues started.
6. After deleting the folder, at one point the entire resource appeared empty - i.e., the files just disappeared. After exiting edit mode, the zip file reappeared. I don't have the exact sequence that caused this.
7. After deleting the folder and then unzipping the original file again, the folder reappears, but when I double click on it to view its contents it takes multiple minutes to show me the folder contents - and it's empty. The unzip created the folder, but the files do not appear to be there. Then, when I back out to the "contents" folder, the original zip file and the folder are gone (the content directory looks empty). Navigating up one folder to the "contents" directory also booted me out of edit mode. When I went back into edit mode, the zip file and folder magically appeared again.
8. After fiddling with this several times now, it appears as though the zip file is there, and the folder is there, but if I try to download the zip file, I get an error saying that it "Failed - Server problem"
9. The resource is now sufficiently broken that Bag Creation fails. I get an error that just says "Sorry. Bag creation failed." I suspect that there is now a complete mismatch between what files are on disk (if any are left on disk) and what files are listed in the Django database.

Given that we have resource types like composite and model instance that may well have hundreds of files in them within a folder hierarchy, we need to make this all a bit more robust.

Since the resource I was testing on now seems to be corrupt, I have shared the original zip file here:

https://drive.google.com/open?id=0B0mUuf2-qdlTX3JVWlZXUkdkUDQ

**Comment by dtarb** [comment]
@hyi could you please look at these from a back end perspective.  @Maurier could you please look at front end aspects of these.

**Comment by Maurier** [comment]
Most of the errors in the front end (if not all of them) could be caused by a bug recently introduced in one of the scripts:

![image](https://user-images.githubusercontent.com/2448568/29138028-8c60f300-7cff-11e7-8dd1-a8e54c7bfd5d.png)

I identified this issue last week. I wrote a fix for it and it was merged recently into our development branch (https://github.com/hydroshare/hydroshare/pull/2264/files#diff-78e89319ca2c700ced648914fc567c2cR1430). I will do some testings in out development branch to assess if the errors are resolved.

**Comment by Maurier** [comment]
As I tested it seems that the errors originate after the unzipping fails. All further operations on the folder will result in bad gateway or time out errors:

Unzipping this file takes a very long time (5 minutes or more) and returns the following error:
![image](https://user-images.githubusercontent.com/2448568/29141735-aa7908d4-7d0c-11e7-8138-b436e8090de9.png)

Trying to open this folder results in a time out error:
![image](https://user-images.githubusercontent.com/2448568/29142617-b514be0c-7d0f-11e7-838f-d007f0541e10.png)

After this point the breadcrumb links are corrupted.

Deleting the folder also results in time out
![image](https://user-images.githubusercontent.com/2448568/29143051-15f94ce6-7d11-11e7-955e-d8b26aa169ec.png)

This error is reproducible using the zip file @horsburgh includes above. The unzip operation should be looked at from the back end. I have deployed our develop branch at https://playground.hydroshare.org for testing.

**Comment by hyi** [comment]
I remember I had investigated the backend unzip operation for a different issue and I remember the conclusion was that unzipping operation from iRODS was very fast, with the majority of time being spent on registering the large amount of unzipped files in Django. I remember I had some timing done to illustrate this. I remember I improved the Django registration for large amount of unzipped file a bit but not by much. One option is to delegate this registration of large amount of unzipped files to a celery task to be done asynchronously to avoid time out issue. The problem is that now we use Django as the point of truth for files, so before files are registered in Django, the user interface will not be able to see files. So front end user interface may need some work to display informational message to notify users that unzipping is in process until it is done. I can work on this (may need a bit of help from @Maurier on the user interface front end) when I have time. In the meantime, we can use this issue to keep track of other ideas for addressing this timeout performance issue.

**Comment by horsburgh** [comment]
Here is where the rubber meets the road in terms of ground truth. Based on what I have read in this issue, we have a user interface function (unzipping) that unzips the files to disk but then the process fails to create all of the correct file references in Django. This leaves the resource in an inconsistent and **broken** state - which is a MAJOR issue. File operations are fundamental to our system and MUST be handled correctly. If Django is to be the ground-truth for files, then we can't afford for this process to be broken.

@hyi - your solution may be reasonable, but I'm curious whether there are any other options. @alvacouch - do you have any ideas here?

**Comment by ChristinaB** [comment]
@alvacouch 
Please add datestamp column in the resource contents.  This helps with sorting and versioning and verifying.  Thanks!

It personally is an issue for me that I don't know when a file was last uploaded. And then I got an email from a colleague related to Collections, their issue was that the generic resource has been updated, but the date listed for the resource in the Collection is not.  I think users are accustomed to checking the datestamp of files as a quality control check to be sure that a file was uploaded correctly.  Is the Collection pointer part of a different issue?

**Comment by alvacouch** [comment]
@horsburgh @ChristinaB @dtarb @Maurier @aphelionz @martinseul 

Please see [this document in google drive](https://docs.google.com/document/d/1BCyvQjInF9zkosvaGokG5hFkcMAhAJZ3aI86KR3I54I/edit#) for my detailed thoughts on this problem. 

I am prepared to switch the ground truth to iRODS from Django, but there are some downsides that affect usability. Pay particular attention to the two options at the bottom of the file listed above. 

The key here is that no matter how I handle things, there is one operation in HydroShare -- generating `sciencemetadata.xml` -- that requires that Django and iRODS be in sync. This is true for all resource types, but particularly true for composite resources. All ways in which I can improve performance require some asynchronous operations that return control to the user before completing. The question is precisely when to insure synchronization in order to do the one thing synchronization is required for, which is `sciencemetadata.xml`. 

1. As a queued operation that marks the resource as unusable while it completes. 
2. As part of the creation of `sciencemetadata.xml` itself, which means that the resource remains usable but you can't fetch the science metadata until the files synchronize. So, when you change files, "create bag" and "fetch sciencemetadata.xml" become asynchronous operations that return "in progress" until they're done.

I prefer the second option.

**Comment by horsburgh** [comment]
@alvacouch - I think option 2 can be workable if the right messages are provided to the user so they know what to expect.

**Comment by hyi** [comment]
@alvacouch @horsburgh I think option 2 is in line with my thinking when I proposed to make the time-consuming file operation to run asynchronously as celery tasks. In a way, it is kind of like async bag download operation where the wait/in progress message shows before the bag creation is done by async celery task.

**Comment by dtarb** [comment]
@alvacouch I added some comment text to the [document in google drive](https://docs.google.com/document/d/1BCyvQjInF9zkosvaGokG5hFkcMAhAJZ3aI86KR3I54I/edit#) that you mentioned.  Nothing profound, just clarifying some of the raster handling details.  I am supportive of the approach to switch ground truth to iRODS and also favor option 2 above.  It is better to only disable the parts that depend on science metadata, not the whole resource, while the science metadata is being generated.

**Comment by alvacouch** [comment]
@dtarb @horsburgh There is one downside to option 2 that I didn't mention, which is that the science metadata will not be consistently available when interacting with files in the file browser. This is again a speed issue. E.g., if the user uploads 100 new files, the file browser will not immediately try to harvest their metadata, but instead, will provide what information it has at the time of the request. I hope this is alright.

**Comment by alvacouch** [comment]
@dtarb @horsburgh @hyi @aphelionz @martinseul @ChristinaB @Maurier See the end of the [document in google drive ](https://docs.google.com/document/d/1BCyvQjInF9zkosvaGokG5hFkcMAhAJZ3aI86KR3I54I/edit#) for some comments on how the UI would handle lazy synchronization.

**Comment by dtarb** [comment]
@alvacouch I read the comments at the end of the document in google drive and I do not really see them as a problem.  There is currently a "please wait for the bag to be created message" when a user clicks to download the bag for a resource where the bag has not been created.  This is accepted current behavior.  I do not see a need to change the interface behavior.

Also, there is no web browser UI element for accessing resourcemetadata.xml (I assume this is the file you are referring to - there is no sciencemetadata.xml).  This is only accessed by the REST API.  I do not see too much of a problem if an API call gets a stale file if an upload is in progress, because the behavior would be the same as if the call was made before the upload or file change was made.  API content consumer programs need to recognize that metadata content is dynamically and may require refreshing.

With respect to harvesting metadata and enabling editing of human specified metadata, I think that metadata should be harvested once file all file operations end.  This needs to be resource type specific.

1. Generic resource.  This may hold many files, but metadata does not need to be harvested from files.
2. Composite resource.  This may hold many files but metadata does not need to be harvested from files at upload.  Metadata is only harvested from files at the time they are set to a supported type.
3.  Geographic raster resource.  This may hold multiple files, but may only be created by uploading single files, either a single .tif file or a single .zip file with multiple files organized using the VRT format.  Harvesting can proceed once the single file upload is complete and zip file unzipped, if a zip file.
4.  Geographic feature resource.  There is a small upper bound on the number of files that can be in a shapefile (on the order of 10).  This is probably the only tricky one because harvesting metadata has to wait for all file uploads to complete.  We could simplify this by requiring single zip file uploads only (at very limited inconvenience to user).  
5.  Model instance.  I do not know the data model for this, so we would need input from @goodalljl

**Comment by alvacouch** [comment]
@dtarb @goodalljl @horsburgh The way that I envision this working is that there'll be a validation step after upload that checks whether the set of uploaded files is appropriate. If the set of files is inappropriate, the UI and/or REST API will inform the user. It will not ever delete files. The user will have to delete any extraneous files before the resource will be considered valid. 

* This is a departure from the current UI, which instead rejects non-conforming uploads. 
* This modification is necessary, however, in order to enable use of 3rd party storage including DropBox and Box. In those cases, there is no "upload" step to police.

**Comment by horsburgh** [comment]
@alvacouch and @hyi - I'm working on a list of issues for @Maurier to work on. This is one that is assigned to him, but I'm not sure what he needs to contribute here yet. He wrote the file interface and so needs to implement any front end changes for this issue, but did we come to any sort of resolution on what needs to be done here?

**Comment by hyi** [comment]
@horsburgh I am not very clear on specifics for this issue either based on the discussion thread, and will defer to @alvacouch to clarify. There seems to still be some designs being involved and not finalized yet, so I am not sure where to start with this issue either other than we make iRODS as the ground truth for files and make time-consuming operations (I thought it is the big loop for registering unzipped files back to Django, but if my memory serves, @alvacouch thought it was the iRODS "ils" that took too much time) async operations so that they don't block the main web thread, in the similar way as I did for bag download.

**Comment by dtarb** [comment]
This issue is a bit big and complex to be practically workable.  Decided to follow the strategy
Assign to @sblack-usu to investigate after 1.16 is deployed and break into actionable pieces.  Develop an initial strategy for the team meeting in August.

**Comment by sblack-usu** [comment]
I'm looking through the code and I'm not seeing a compelling reason for the ResourceFile object outside of registering a logical file type in the composite resource.  For example, the method called by the file browser does a list directory to get a list of all files in that directory.  It then loops through each of those files and nests a loop of all ResourceFile objects for that resource id.  The only thing it extracts out of the found ResourceFile object is the primary key... which seems to be used only for update/deletion methods of the file.  I don't see a reason why we couldn't just use the path for those operations.  It would drop the cost of listing a folder from n^2 to n.

The only time it seems to be helpful is with the introduction of the composite resource and setting the logical file type.  I would propose only using this model when a file has a logical file type... or perhaps we might be able to work something out with metadata on the filesystem.  Thoughts, concerns or anything I'm not seeing here?

**Comment by cuahsimarko** [comment]
This work ticket would be more actionable if it were updated for the last year of fixes and broken into specific, outstanding issues, perhaps high qty/large files.

**Comment by eclark-cuahsi** [comment]
@martinseul please review and follow up.

**Comment by martinseul** [comment]
@alvacouch @sblack-usu can we close this ticket as it's a fairly old, some or all aspects have been addressed or don't apply anymore. If any aspects still need to be addressed I'd suggest opening a new ticket? Please advise or close.
